### PR TITLE
Dune upgrade V1 -> V2

### PR DESCRIPTION
**Pull request by voodoos**
This PR aims to track the ongoing development of a new procedure that would help users to upgrade their projects to `dune_lang >=2.0`. The hope is that a vast majority of projects can be migrated in a mostly automated way. After the upgrade users will be prompted with a list of hard-to-migrate things that they should check manually.

### Doc updates
- [x] `self_build_stubs_archive` does not exists since 2.0
- [x] `c_names` / `c_flags` / etc do not exist since 2.0

### Breaking changes that should be automatically handled
- [x] *Get rid of ad-hoc rules for guessing the version. Dune now only
  relies on the version written in the `dune-project` file and no
  longer read `VERSION` or similar files.*
- [x] *The `action` field in the `alias` stanza is not available starting `lang dune
  2.0`. The `alias` field in the `rule` stanza is a replacement.*
- [x] *Set up formatting rules by default. They can be configured through a new
  `(formatting)` stanza in `dune-project`.*
  💡 The upgrader adds `(formatting disabled)` when the project was not formatted. It adds `(formatted (enabled_for ...))` when needed.
- [x] *Change the default `modes` field of executables to `(mode exe)`. If
  one wants to build a bytecode program, it now needs to be explicitly
  requested via `(modes byte exe)`. *
  💡 The upgrader adds `(modes byte exe)` to executables without explicit mode.
- [x] *Error out when a `preprocessor_deps` field is present but not
  `preprocess` field is. It is a warning with Dune 1.x projects.*
  💡 The upgrader removes `preprocessor_deps` field when there is no `preprocess` field or if the chosen preprocess is `no_preprocessing`
- [x] *`no_keep_locs` is a no-op for projects that use `lang dune` older than 2.0. In
  projects where the language is at least `2.0`, the field is now forbidden.*
- [x] *c_flags, c_names and cxx_names are now supported in executable and executables stanzas.
 Note: this feature has been subsequently extended into a separate foreign_stubs field.*
💡 In libraries and executables stanzas, the upgrader replaces the fields `c(xx)_names` and `c(xx)_flags` by an appropriate `foreign_stubs` field.

### Breaking changes that the user should be aware of
- ~~*Drop support for `jbuild` and `jbuild-ignore` files.*~~
- *mli only modules must now be explicitly declared. This was previously a
  warning and is now an error. *
- `self_build_stubs_archive` was deleted in version 2.0 of the dune language
- *Stop installing the `ocaml-syntax-shims` binary. In order to use
  `future_syntax`, one now need to depend on the `ocaml-syntax-shims`
  package.*
- *Actions which introduce targets where new targets are forbidden (e.g.
  preprocessing) are now an error instead of a warning.*
- *Stop installing the `ocaml-syntax-shims` binary. In order to use
  `future_syntax`, one now need to depend on the `ocaml-syntax-shims`
  package.*
- *Do not put the `<package>.install` files in the source tree unless `-p` or
  `--promote-install-files` is passed on the command line*
- *Library names are now validated in a strict fashion. Previously, invalid names
  would be allowed for unwrapped libraries*
- *Stricter validation of file names in `select`. The file names of conditional
  sources must match the prefix and the extension of the resultant filename.*
- *Modules filtered out from the module list via the Ordered Set Language must
  now be actual modules.*
- *Stub names are no longer allowed relative paths. This was previously a warning
  and is now an error.*
- *In `(diff? x y)` action, require `x` to exist and register a
  dependency on that file.*

### Non-breaking changes
- The rest of the [changelog 2.0.0 section](https://github.com/ocaml/dune/blob/master/CHANGES.md#200-20112019).

### CODE REVIEW

**Comment by diml** [comment]
> Drop support for jbuild and jbuild-ignore files

I don't think we need to do anything about that. dune >= 2.0.0 binaries cannot read jbuild files at all, whether we are in a dune 1 or dune 2 project.

**Comment by voodoos** [comment]
I have updated the list of breaking changes that the upgrader should take into account (first post).
@diml, @nojb, @rgrinberg , @emillon  do you see anything missing ?

**Comment by diml** [comment]
That looks all good to me! I can't think of anything else. Thanks!

**Comment by voodoos** [comment]
After performing some testing I encountered this change:

> Error: 'c_names' was deleted in version 2.0 of the dune language. 
> Use the `(foreign_stubs ...)` field instead.

Which was not very clear in the changelog:

> c_flags, c_names and cxx_names are now supported in executable and executables stanzas. 
> (#2562, @nojb) Note: this feature has been subsequently extended into a separate 
> foreign_stubs field. (#2659, RFC #2650, @snowleopard)

I will add that to the upgrader's work, and correct the doc/changelog.

**Comment by emillon** [comment]
That's great! With `(foreign_stubs)` that'll cover most of the uses.

**Comment by voodoos** [comment]
Ok, I think this PR is ready for review.

It's a bit of a mouthful, so here are a few pointers:

Every thing is in `upgrader.ml`. I initially splitted the upgrader in three files, "common", "v1" and "v2", but decided to merge them back at the end.

- `module Common` contains code which is common to both upgraders or that should be useful for a future v2 -> v3 upgrader.
-  `module V1` contains the upgrader from jbuild to dune v1.
- `module V2` contains the novel upgrader from dune v1 to v2.
- The toplevel `upgrade` function makes the list of folders to upgrade and to which version, then calls the corresponding upgraders. The project version "detection" is happening in the `detect_project_version` function and may still be a bit fragile.

To update an old jbuild project to v2, the upgrader should be run twice.

**Comment by rgrinberg** [comment]
> To update an old jbuild project to v2, the upgrader should be run twice.

Is it difficult to make the tool sequence the upgrades automatically? it's not so important and I think that it probably makes sense for users to inspect each upgrade individually, but for some duniverse like use cases where you have a massive workspace of 3rd party packages, it would be annoying to run `$ dune upgrade` a few times.

**Comment by emillon** [comment]
Yes I agree about this point, in particular that ensures that `dune upgrade` is idempotent (running it twice is equivalent to running it once).

**Comment by voodoos** [comment]
> Is it difficult to make the tool sequence the upgrades automatically? 

Ok so I tried to make the updater run twice in two ways:
- Juste make it call itself recursively at the end of the first ugprade if v0-v1 upgrades where detected.
- `Schedule` two upgrades at the `bin/upgrade.ml` level

In both cases it looks like the `File_tree` is not refreshed: the second upgrades tries to perform v0->v1 upgrades again but fails because the `jbuild` files have already been translated and deleted.

Any idea of how I could manage it ?

**Comment by rgrinberg** [comment]
Indeed the File_tree is not refreshed. However, this isn’t something that is too tough to fix.

If you look at File_tree.Settings.get, you will see that it depends on the current run. I suppose that we’d like for this to depend on a “ref” like value that we can control to reset the file tree after the upgrade finishes.

The only trick here is that we need this ref to respect memoization.

I’ve copied Arseniy. He can advise us if this is the best way to solve this problem.

**Comment by voodoos** [comment]
Thanks, `Memo` has a `reset` function.
If I call this function before the new upgrading attempt it does work !

Is that a bad way to do it ?

**Comment by rgrinberg** [comment]
It’s not ideal but it should work as well in this case. The only negatives is that it could be confusing to future readers, and that it could possibly be slow things down. I think it’s fine to use here if you leave a comment why it’s necessary.

**Comment by aalekseyev** [comment]
I think calling `reset` is reasonable in this case. The incremental computation is not fully implemented yet and untested so we shouldn't try to make it work in this PR.

**Comment by emillon** [comment]
Hello,

Just checking in: what's the status of this? If it's doing the upgrades in an idempotent way I think it's good to go after a final round of review.

**Comment by voodoos** [comment]
I added a test illustrating the idempotency of the upgrade procedure.
I also rebased on master. It should be ready for review.

File: src/dune/upgrader.mli [hunk 1]
```diff
@@ -1,4 +1,4 @@
 (** Upgrade projects from jbuilder to Dune *)
 
 (** Upgrade all projects in this file tree *)
-val upgrade : unit -> unit
+val upgrade : ?last:bool -> unit -> unit
```
Comment (emillon) [inline src/dune/upgrader.mli]: It seems that this param is only used in the recursive call. In order not to pollute the interface, what do you think about extracting a method with a (mandatory) param and leave `upgrade` without param?

File: src/dune/upgrader.ml [hunk 2]
```diff
@@ -399,4 +686,21 @@ let upgrade () =
         ];
       List.iter (original_file :: extra_files_to_delete) ~f:(fun p ->
           Path.unlink (Path.source p));
-      Io.write_file (Path.source new_file) contents ~binary:true)
+      Io.write_file (Path.source new_file) contents ~binary:true);
+  if !v1_updates && not last then (
+    (* Run the upgrader again to update new v1 projects to v2 No more than one
+       additionnal upgrade should be needed *)
+    (* We reset thje memoization as a simple way to refresh the File_tree *)
```
Comment (emillon) [inline src/dune/upgrader.ml]:
```suggestion
    (* We reset memoization as a simple way to refresh the File_tree *)
```

File: src/dune/upgrader.ml [hunk 3]
```diff
@@ -399,4 +686,21 @@ let upgrade () =
         ];
       List.iter (original_file :: extra_files_to_delete) ~f:(fun p ->
           Path.unlink (Path.source p));
-      Io.write_file (Path.source new_file) contents ~binary:true)
+      Io.write_file (Path.source new_file) contents ~binary:true);
+  if !v1_updates && not last then (
+    (* Run the upgrader again to update new v1 projects to v2 No more than one
+       additionnal upgrade should be needed *)
```
Comment (emillon) [inline src/dune/upgrader.ml]:
```suggestion
       additional upgrade should be needed *)
```

File: src/dune/upgrader.ml [hunk 4]
```diff
@@ -46,343 +10,666 @@ type rename_and_edit =
...
         ]
+    (* Unexpected *)
+    | _ -> sexps
+
+  let update_project_file todo project =
+    let sexps, comments = read_and_parse (Dune_project.file project) in
+    let v = !Dune_project.default_dune_language_version in
+    let sexps = sexps |> Ast_tools.bump_lang_version v |> update_formatting in
+    let new_file_contents = string_of_sexps sexps comments in
+    (* Printf.eprintf "AAAH: %b\n!" (was_formatted sexps); *)
```
Comment (emillon) [inline src/dune/upgrader.ml]:
```suggestion
```

File: test/blackbox-tests/test-cases/upgrader/partv2/partv1/dune [hunk 5]
```diff
@@ -0,0 +1,37 @@
+(include dune.inc)
+
+(alias
+  (name runtest)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/upgrader/partv2/partv1/dune]: Is this formatting correct? It should be one space

File: src/dune/upgrader.ml (hunk 4 above)
Comment (voodoos) [reply src/dune/upgrader.ml]: hmm

**Comment by emillon** [comment]
once the upgrade type is fixed, formatting is fixed and DCO is added I think we can squash & merge.

**Pull request merged**